### ISSUE TITLE
events should use json formatting

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#2106]
Noticed the guest exec eventlog showed:

```
>flux kvs eventlog get job.active.0000.0062.5400.0000.guest.exec.eventlog
2019-03-25T18:24:10.270071Z init
2019-03-25T18:24:10.273329Z starting
2019-03-25T18:24:10.273366Z running timer=0.000010s
2019-03-25T18:24:10.274506Z complete status=0
2019-03-25T18:24:10.274569Z done
```

should we be using json format for `timer=...` and `status=...`?  I don't believe the exec eventlog is explicitly listed in RFC21, but perhaps should be done for consistency.

**Comment by grondo** [comment]
This is just debug data, I'm not sure the switch to JSON context is worth it here (these entries are simulated job shell events). Unless we want to ponder if RFC 18 should require all KVS eventlog context be in JSON instead of uspecified...

**Comment by chu11** [comment]
@grondo ok, no biggie.  I was thinking about how to implement #2105, but the formatting would be a problem.  We can wait as the exec system unfolds more.  Feel free to close issue if it's something that could disappear in future.

**Comment by grondo** [comment]
Now that I think about it more, you might be right. We've already got a complicated issue with multiple eventlogs per job, it will make the problem slightly more tractable if we keep the context of exec system events in the same format as the main job eventlog.

I had an issue open on defining the required exec system events. We should nail that down as was done in RFC21 and define the required context, keeping it simple at first.

**Comment by garlick** [comment]
> Unless we want to ponder if RFC 18 should require all KVS eventlog context be in JSON instead of uspecified...

Would it simplify our lives if RFC 18 specified that each event is a JSON object that requires `timestamp` and `name` keys and allows other optional keys?  I didn't want to go this route at first because the whole eventlog would not be valid JSON (and valid JSON spread over multiple lines would not be a valid event), but in retrospect those constraints can be specified in the RFC and seem fairly inconsequential.

How this impacts the KVS eventlog API in `kvs_eventlog.h` is sort of intriguing.  Could we represent the entire eventlog as a JSON array and drop the name and context length limitations?  It might be worth exploring how that API could change for the better or even be simplified to the point where we don't need it in the public API.

Or maybe "wait and see" makes more sense right now...

**Comment by chu11** [comment]
> Would it simplify our lives if RFC 18 specified that each event is a JSON object that requires timestamp and name keys and allows other optional keys?

I think it's worth considering.  While making contexts JSON in RFC21 helped us remove a lot of the utility parsing functions in job-manager (#2085), we still have a number of encode/decode functions which could maybe be removed as well.  Or if we want to keep the encode/decode functions for convenience, internally the parsing would be easier.

>  How this impacts the KVS eventlog API in kvs_eventlog.h is sort of intriguing.  Could we represent the entire eventlog as a JSON array and drop the name and context length limitations?

Are you imaging some API function would get the raw contents of the eventlog and then re-make it into a single json array for a user?  Now we can remove the iterating functions in `kvs_eventlog`?

**Comment by grondo** [comment]
In general, I think this change makes a lot of sense for the long run.

**Comment by garlick** [comment]
> Are you imaging some API function would get the raw contents of the eventlog and then re-make it into a single json array for a user? Now we can remove the iterating functions in kvs_eventlog?

Something like that, either with an opaque type and iterators in the public API, or just plain JSON, without iterators, if we decided to take that interface private.

**Comment by chu11** [comment]
I began working on converting all events to use json format today, and the work may less than I originally anticipated.  Since alot of encoding / decoding of events goes through the `libkvs` API, the RFC changes can be abstracted in there.

However, what will likely change is APIs in `libkvs`.  Functions like:

```
int flux_kvs_event_decode (const char *s,
                           double *timestamp,
                           char *name, int name_size,
                           char *context, int context_size);
char *flux_kvs_event_encode (const char *name, const char *context);
char *flux_kvs_event_encode_timestamp (double timestamp,
                                       const char *name,
                                       const char *context);
```

could be tweaked, take a `json_t *o` parameter, have a `pack` equivalent function to support a json object in the context, etc.  I imagine we can get rid of the maximum length `name` and `context` macros as well.  Still waiting to see how things fall out first.

**Comment by garlick** [comment]
Let's avoid exposing a `json_t *o` parameter directly in the API in consideration of bindings and C users with other JSON libraries, but pack/unpack plus a string interface for contexts sounds good.  Were you thinking of having both abstract `flux_kvs_event_t` and `flux_kvs_eventlog_t` types, or...?

**Comment by chu11** [comment]
> Were you thinking of having both abstract flux_kvs_event_t and flux_kvs_eventlog_t types, or...?

I hadn't thought of adding a `flux_kvs_event_t` yet, but that could be in the cards.  Still trying to see how things fall out.

**Comment by grondo** [comment]
Premature since you're not sure how things are falling out, but `flux_kvs_event_t` could be a confusing type name. Unfortunately, it would probably be better to choose something with `eventlog` in the name, e.g. `flux_kvs_eventlog_entry_t`.

**Comment by chu11** [comment]
in #2117 did the first round of changes, which just converts all events/eventlogs to JSON object format.  

The first and most obvious change is to support a `flux_kvs_event_encode_pack()` so that arbitrary contexts can be encoded without constantly needing to build them up them manually.

```C
char *flux_kvs_event_encode_pack (const char *name, const char *fmt, ...);
char *flux_kvs_event_encode_timestamp_pack (double timestamp,
                                       const char *name,
                                       const char *fmt, ...);
```

We'll probably also need a `vpack` equivalent too.

A subtle change will be to make it so that if the user inputs a context of `""` (i.e. empty string), that'll now be an error.  Before, when a context was an arbitrary string, such an input made sense.  But now that a context is a JSON object, it makes less sense to allow that.  There will need to be code adjustments for this in a few places (notably `job-exec`).

Now, `flux_kvs_event_decode_unpack()` is something that is also necessary, but that's a bigger can of worms.  The naive new function prototype:

```C
int flux_kvs_event_decode_unpack (const char *s,
    double *timestamp,
    char *name, int name_size,
    const char *fmt, ...);
```

But now we have multiple problems.  A) should we remove the name_size limit?  Probably.  B) what about storage of anything returned in the varargs part?  B/c the common case is that we're going to be passing in the string returned from a KVS lookup.  Should we return an opaque type?

```C
flux_kvs_eventlog_entry_t *flux_kvs_event_decode_unpack (const char *s,
    double *timestamp,
    char **name,
    const char *fmt, ...);
```

This can deal with all storage issues and the user just destroys `flux_kvs_eventlog_entry_t` when they're done?  There could also be getters for this opaque type?

```C
flux_kvs_eventlog_entry_t *flux_kvs_event_decode (const char *s);
double flux_kvs_eventlog_entry_timestamp (flux_kvs_eventlog_entry_t *e);
const char *flux_kvs_eventlog_entry_name (flux_kvs_eventlog_entry_t *e);
const char *flux_kvs_eventlog_entry_context (flux_kvs_eventlog_entry_t *e);
int flux_kvs_eventlog_entry_context_unpack (flux_kvs_eventlog_entry_t *e, const char *fmt, ...);
```

But this suddenly seems to be getting long ... :-(  I dunno.

Another subtlety is, should `flux_kvs_event_decode()` require a newline at the end of the event?  Right now it does, but it sort of shouldn't matter once to get to this point.  It shouldn't be relevant if parsing code stripped the newline or not.

Now onto the `struct flux_kvs_eventlog` functions:

```C
struct flux_kvs_eventlog *flux_kvs_eventlog_create (void);
void flux_kvs_eventlog_destroy (struct flux_kvs_eventlog *eventlog);
char *flux_kvs_eventlog_encode (const struct flux_kvs_eventlog *eventlog);
struct flux_kvs_eventlog *flux_kvs_eventlog_decode (const char *s);
int flux_kvs_eventlog_append (struct flux_kvs_eventlog *eventlog, const char *s);
const char *flux_kvs_eventlog_first (struct flux_kvs_eventlog *eventlog);
const char *flux_kvs_eventlog_next (struct flux_kvs_eventlog *eventlog);
```

`flux_kvs_eventlog_encode()` isn't used anywhere except in tests.  Will there be a use case for this longer out?  Where users will manipulate an eventlog and want to re-write it out back to the KVS?  I just can't see a use case.

`flux_kvs_eventlog_decode()` is only used once in `job-manager`.

`flux_kvs_eventlog_append()` is only used three times.  Once in `flux_kvs_eventlog_decode()`, once in `flux-job` and once in `flux-kvs`.  Of the three uses, only in `flux-kvs` do we actually "append".  The other cases are "load an entire eventlog so we can iterate over it".

If we remove `--watch` from `flux kvs eventlog get` (which do we need it anymore?), then effectively the only use for all of these functions is "load the entire eventlog so we can iterate over it".

We could simplify the API to

```C
flux_kvs_eventlog_t *flux_kvs_eventlog_create (const char *eventlog_string);
void flux_kvs_eventlog_destroy (flux_kvs_eventlog_t *eventlog);
const char *flux_kvs_eventlog_first (flux_kvs_eventlog_t *eventlog);
const char *flux_kvs_eventlog_next (flux_kvs_eventlog_t *eventlog);
OR
flux_kvs_eventlog_entry_t *flux_kvs_eventlog_first (flux_kvs_eventlog_t *eventlog);
flux_kvs_eventlog_entry_t *flux_kvs_eventlog_next (flux_kvs_eventlog_t *eventlog);
```

?

I'm going to start on `flux_kvs_event_encode_pack()` for now, b/c that's an obvious need.  The rest is sort of debatable.

**Comment by grondo** [comment]
> Now, flux_kvs_event_decode_unpack() is something that is also necessary, but that's a bigger can of worms.

Maybe we just treat an eventlog entry as similar to `json_t *` and have a simpler convenience unpack?

```C
int flux_kvs_eventlog_entry_unpack (flux_kvs_eventlog_entry_t *e, const char *fmt, ...);
```

Where the guaranteed keys would be `name` and `timestamp`?
(actually, maybe that is not any better than your proposal)

My preference in any event is that we borrow the reference for `name` and return it as `const char *` and do not deal with the name size parameter anymore.

Also, @garlick and I were talking this morning and while we definitely shouldn't export interfaces using `json_t *` via the public flux API, perhaps all this code will be internal only and therefore a `json_t *` might be ok. Would that greatly simplify your work here?

**Comment by chu11** [comment]
> int flux_kvs_eventlog_entry_unpack (flux_kvs_eventlog_entry_t *e, const char *fmt, ...);

I didn't even think about this, I think it would be easier.  If the caller needs `name` or `timestamp` they can just get it that way.

> perhaps all this code will be internal only and therefore a json_t * might be ok. Would that greatly simplify your work here?

I think so.  I imagined created a function like `json_t *flux_eventlog_to_json_array()`, and bam, it's jansson functions the rest of the way for reading / decoding eventlogs.  We might keep `flux_kvs_event_decode()` in some form for convenience, but that's about it.

**Comment by garlick** [comment]
I would be OK with going internal only for now, and if we have a use case (e.g. a scheduler having its own eventlog?) later on developing an external wrapper for it.

If we're going with internal only, I wonder if this belongs in libkvs?  Should we put it in libjob or maybe a libeventlog?

**Comment by garlick** [comment]
@chu11 would it be helpful if I proposed an internal API based on exposed `json_t` or do you have something in mind already?

**Comment by chu11** [comment]
With the introduction of `pack` and `vpack` variants, I now have:

```C
char *flux_kvs_event_encode (const char *name, const char *context);
char *flux_kvs_event_encode_timestamp (double timestamp,
                                       const char *name,
                                       const char *context);
char *flux_kvs_event_encode_pack (const char *name,
                                  const char *context_fmt,
                                  ...);
char *flux_kvs_event_encode_timestamp_pack (double timestamp,
                                            const char *name,
                                            const char *context_fmt,
                                            ...);
char *flux_kvs_event_encode_vpack (const char *name,
                                   const char *context_fmt,
                                   va_list ap);
char *flux_kvs_event_encode_timestamp_vpack (double timestamp,
                                             const char *name,
                                             const char *context_fmt,
                                             va_list ap);
```

Which seems like an unnecessary permutation of functions.  The original non-pack/non-vpack functions can probably eventually go away.

Perhaps make a change so that a timestamp < 0 means "get the current time"?  So we can remove a permutation?

**Comment by garlick** [comment]
> Perhaps make a change so that a timestamp < 0 means "get the current time"?

RFC 18 says timestamp must be > 0, so 0 would be OK to use there IMHO.

Or maybe rather than create+encode, and assuming an "internal" API, and assuming _event_ is confusing as discussed, go with something like
```c
json_t *eventlog_entry_create (double timestamp, const char *name, const char *context);
json_t *eventlog_entry_pack (double timestamp, const char *name, const char *fmt, ...);
```

**Comment by garlick** [comment]
For encoding/decoding eventlogs or eventlog fragments represented as JSON arrays:
```c
json_t *eventlog_decode (const char *s);
char *eventlog_encode (json_t *eventlog);
```
(use json_array_foreach(), json_array_extend(), json_array_append(), etc. to manipulate)

**Comment by chu11** [comment]
> json_t *eventlog_entry_create (double timestamp, const char *name, const char *context);

The majority of the time encode is used, it's used to create a new event entry in the KVS.  So I think returning `char *` in those functions is ok.

> If we're going with internal only, I wonder if this belongs in libkvs? Should we put it in libjob or maybe a libeventlog?

I'm thinking a new `libeventlong` library.

**Comment by chu11** [comment]
just re-pushed #2117 with a new `libeventlog` convenience library.  Here's the header of the new functions I added.  

```
int eventlog_entry_parse (json_t *event,
                          double *timestamp,
                          const char **name,
                          json_t **context);

json_t *eventlog_decode (const char *s);

json_t *eventlog_entry_decode (const char *event);

char *eventlog_entry_encode (double timestamp, const char *name,
                             const char *context);

char *eventlog_entry_encode_vpack (double timestamp,
                                   const char *name,
                                   const char *context_fmt,
                                   va_list ap);

char *eventlog_entry_encode_pack (double timestamp,
                                  const char *name,
                                  const char *context_fmt,
                                  ...);
```

I still need to add unit tests, header comments, remove `flux_kvs_eventlog`, etc, but hopefully its obvious what these do.  The only oddball one is `eventlog_entry_parse()`.  I was calling `json_unpack()` in a certain pattern so often, just better to make a common function.

It's nice to be able to work with `json_t` everywhere, removing the need for the `flux_kvs_eventlog_*` calls.  And the `pack` & `vpack` functions are much better than manually creating JSON contexts everywhere.  And being able to use `json_unpack()` on the contexts is nice too.

The one oddball area is in the `job-manager`.  In most cases, when an eventlog entry is created/encoded, it's just to write it to the KVS.  So returning a `char *` from all of the `eventlog_entry_encode()` functions makes sense.

But in `job-manager`, after an event is created, several times you call `event_job_update()`, which wants to immediately parse the event.  So what would be best?  For the time being, I created a `event_job_update_strevent()` which takes a string event and converts to a `json_t`.  The main `event_job_update()` takes a `json_t *` event.

I contemplated doing `char *eventlog_entry_encode_pack (..., json_t **event)`, where the caller can get both a `json_t` and `char *` of the event.  But that seemed bad.  Not sure if there's a refactoring that could be done in `job-manager` to make this better, but nothing obvious came to mind.

**Comment by garlick** [comment]
Apologies for not getting back to this.  Sounds reasonable based on your explanation, but I'll check it out and poke at it a bit.

**Comment by chu11** [comment]
Thinking about it this weekend.  I think this pattern suggested by @garlick

```
json_t *eventlog_entry_(v)pack (double timestamp, const char *name, const char *fmt, ...);
char *eventlog_entry_encode (json_t *entry);
```

Might work a tad better.   I'll experiment and see if it ends up nicer.

**Comment by chu11** [comment]
tweaked the `libeventlog` library to the following:

```C
/* convenience function to extract timestamp, name, and optional
 * context from an event entry */
int eventlog_entry_parse (json_t *entry,
                          double *timestamp,
                          const char **name,
                          json_t **context);

/* decode an eventlog into an json array of event objects */
json_t *eventlog_decode (const char *s);

/* decode a single eventlog entry into a json object */
json_t *eventlog_entry_decode (const char *entry);

/* build an eventlog entry, context must be a json object */
json_t *eventlog_entry (double timestamp, const char *name,
                        const char *context);

/* build an eventlog entry using jansson style pack, context must be
 * a json object */
json_t *eventlog_entry_pack (double timestamp,
                             const char *name,
                             const char *context_fmt,
                             ...);
json_t *eventlog_entry_vpack (double timestamp,
                              const char *name,
                              const char *context_fmt,
                              va_list ap);

char *eventlog_entry_encode (json_t *entry);
```

So the normal case for encoding is now:

```
json_t *entry = eventlog_entry() / eventlog_entry_pack () / eventlog_entry_vpack();
char *entrystr = eventlog_entry_encode (entry);
```

I dislike the extra call to `eventlog_entry_encode()`, b/c its unnecessary at times.  But overall, it appears to be a net win code wise b/c of how the `job-manager` desires to create entries and parse them before writing them to the KVS.

**Comment by garlick** [comment]
LGTM!

Might need header comments about about how context can be NULL in `eventlog_entry()` (should that be `eventlog_entry_create()`?)  and how `context_fmt` should not be NULL in the pack variants.

I assume it works out that we need the pack/vpack, as opposed to just an `eventlog_entry{_create}()`  that accepts a `json_t *context`?

Should we add a function that gets the current wall clock in seconds-since-epoch format as specified by the RFC?

**Comment by chu11** [comment]
> Might need header comments about about how context can be NULL in eventlog_entry() (should that be eventlog_entry_create()?) 

Ahh, good idea.  I like `eventlog_entry_create()`, it'll be easier to grep for that :-)  I should also document that `timestamp` = `0.0` means get the current time.

> how context_fmt should not be NULL in the pack variants.

Right now I allow `context_fmt` to be NULL.  Do you think that's a mistake?

> I assume it works out that we need the pack/vpack, as opposed to just an eventlog_entry{_create}() that accepts a json_t *context?

Yeah, it's simply a convenience.  Instead of the pattern:

```
context = json_pack (...);
entry = eventlog_entry_create (timestamp, name, context);
```

We can do it within 1 call.

> Should we add a function that gets the current wall clock in seconds-since-epoch format as specified by the RFC?

Now that you mention it, that would be convenient.  I think that code is cut & pasted in a few places.

**Comment by garlick** [comment]
> Right now I allow context_fmt to be NULL. Do you think that's a mistake?

Yeah, users should be encouraged to use the non varargs function in that case IMHO.

**Comment by chu11** [comment]
> Now that you mention it, that would be convenient. I think that code is cut & pasted in a few places.

Actually, this is only in 1 place.  And now that I think about it, the user should just get the timestamp from the eventlog entry that was created.

**Comment by chu11** [comment]
> Yeah, users should be encouraged to use the non varargs function in that case IMHO.

Hmmm.  It ends up allowing NULL `context_fmt` is quite convenient.  Now that contexts are JSON objects, the function:

```
json_t *eventlog_entry_create (double timestamp, const char *name, const char *context);
```

Really only has 1 use case.  It's in `flux-kvs` where the user can pass an arbitrary context string on the command line.

Everywhere else, you'll never pass non-NULL to `eventlog_entry_create()`.  So the use is either

```
json_t *eventlog_entry_create (timestamp, name, NULL):
```

or

```
json_t *eventlog_entry_{v}pack (timestamp, name, "{ ... }", ...);
```

So by not allowing, NULL in `eventlog_entry_{v}pack()` there will always be an extra set of functions to handle the NULL context case.

**Comment by garlick** [comment]
I guess there is nothing wrong with allowing a NULL.  I got annoyed before with needing to do this:
```c
if (errstr)
    flux_respond_error (h, msg, "%s", errstr);
else
    flux_respond_error (h, msg, NULL);
```
but in this case we have the other function for passing a string so that won't happen.

**Comment by chu11** [comment]
Fixed in #2128
